ulog.cpp

```diff
@@ -0,0 +1,316 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "ulog.h"
+
+#include <assert.h>
+
+#include "log.h"
+#include "util.h"
+
+#define ULOG_HEADER_SIZE 16
+#define ULOG_MAGIC                               \
+    {                                            \
+        0x55, 0x4C, 0x6F, 0x67, 0x01, 0x12, 0x35 \
+    }
+
+#define NO_FIRST_MSG_OFFSET 255
+
+struct _packed_ ulog_msg_header {
+    uint16_t msg_size;
+    uint8_t msg_type;
+};
+
+ULog::ULog(const char *logs_dir, Timeout_Manager *timeout_manager)
+    : Endpoint{"ULog", false}
+{
+    assert(timeout_manager);
+    assert(logs_dir);
+
+    _timeout_manager = timeout_manager;
+    _logs_dir = strdup(logs_dir);
+}
+
+ULog::~ULog()
+{
+    free((void *)_logs_dir);
+}
+
+static bool _ulog_timeout_cb(void *data)
+{
+    ULog *ulog = static_cast<ULog *>(data);
+    return ulog->logging_start_timeout();
+}
+
+bool ULog::logging_start_timeout()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_START;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = _target_system_id;
+
+    mavlink_msg_command_long_encode(_system_id, MAV_COMP_ID_ALL, &msg, &cmd);
+    _send_msg(&msg, _target_system_id);
+
+    return true;
+}
+
+bool ULog::start()
+{
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    char path[1024];
+
+    snprintf(path, sizeof(path), "%s/%i-%02i-%02i_%02i-%02i-%02i.ulg", _logs_dir,
+             timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday, timeinfo->tm_hour,
+             timeinfo->tm_min, timeinfo->tm_sec);
+    _file = fopen(path, "wb");
+    if (!_file) {
+        log_error_errno(errno, "Unable to open ULog file(%s): (%m)", path);
+        return false;
+    }
+
+    _timeout = _timeout_manager->add_timeout(MSEC_PER_SEC, _ulog_timeout_cb, this);
+    null_check(_timeout, false);
+
+    _waiting_header = true;
+    _waiting_first_msg_offset = false;
+    _expected_seq = 0;
+    _buffer_len = 0;
+    _system_id = SYSTEM_ID;
+
+    log_info("Logging system_id=%u on %s", _system_id, path);
+    return true;
+}
+
+void ULog::stop()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    if (!_file) {
+        log_error("ULog not started");
+        return;
+    }
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_STOP;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = _target_system_id;
+
+    mavlink_msg_command_long_encode(_system_id, MAV_COMP_ID_ALL, &msg, &cmd);
+    _send_msg(&msg, _target_system_id);
+
+    if (_timeout) {
+        _timeout_manager->del_timeout(_timeout);
+        _timeout = nullptr;
+    }
+
+    fflush(_file);
```

Comment:
Unless accounting for errors - which is not the case, no need to `fflush()` before `fclose()`, as the later already performs the first.